File: cpp/examples/sampler_unprojected.cpp
// Example: unprojected CF state (PsiUnproj) sampled with Sherman-Morrison inverse updates.
//
// Phase 1 thermalizes with the generic logdet-based driver (tunes σ); phase 2 samples with
// the O(N) ratio + O(N²) rank-1 inverse update.
//
// Build (from cpp/): cmake -B build -DCMAKE_PREFIX_PATH=/opt/homebrew && cmake --build build
// Run:               ./build/sampler_unprojected [N] [n] [p] [num_therm] [num_steps] [seed]
#include <cfsonsphere/cfsonsphere.hpp>

#include <cstdio>
#include <cstdlib>
#include <vector>

using namespace cfs;

int main(int argc, char** argv) {
    const int N = argc > 1 ? std::atoi(argv[1]) : 9;
    const int n = argc > 2 ? std::atoi(argv[2]) : 3;
    const int p = argc > 3 ? std::atoi(argv[3]) : 1;
    const long num_therm = argc > 4 ? std::atol(argv[4]) : 200000;
    const long num_steps = argc > 5 ? std::atol(argv[5]) : 1000000;
    const std::uint64_t seed = argc > 6 ? std::strtoull(argv[6], nullptr, 10) : 1;

    auto [twoQ, lm] = cf_ground_state_lm(N, n, p);
    const int Nsys = static_cast<int>(lm.size());

    PsiUnproj cur(twoQ, p, Nsys, lm);
    PsiUnproj nxt(twoQ, p, Nsys, lm);

    auto logpdf = [](const PsiUnproj& psi) {
        return 2.0 * (log_det(psi.slater_det) + psi.jastrow_factor_log).real();
    };

    Rng rng(seed);
    Eigen::VectorXd th, ph;
    rand_theta_phi(rng, Nsys, th, ph);
    Eigen::VectorXd th_n = th, ph_n = ph;

    // Phase 1: thermalize + tune σ with the generic (logdet) driver.
    double sigma = M_PI / std::sqrt(12.0);
    auto [it, sigma_t, dt_therm, acc_therm] =
        gibbs_thermalization(rng, cur, nxt, th, ph, th_n, ph_n, sigma, logpdf, num_therm);
    sigma = sigma_t;
    std::printf("thermalization: acceptance=%.3f  sigma=%.4f  dt=%.2fs\n", acc_therm, sigma, dt_therm);

    // Phase 2: Sherman-Morrison-accelerated sampling.
    initialize_inverse(cur);
    Eigen::VectorXcd temp(Nsys);

    const int n_theta = 200;
    std::vector<double> theta_mesh(n_theta);
    for (int i = 0; i < n_theta; ++i) theta_mesh[i] = std::acos(1.0 - 2.0 * i / (n_theta - 1));
    std::vector<double> density(n_theta - 1, 0.0);

    long naccept = 0;
    for (long mc = 1; mc <= num_steps; ++mc) {
        auto pr = proposal(rng, th(it), ph(it), sigma);
        nxt.update(pr.first, pr.second, it);
        const cdouble dr = slater_det_ratio(cur, nxt, it);
        const double dlog =
            2.0 * (std::log(dr) + nxt.jastrow_factor_log - cur.jastrow_factor_log).real();
        if (dlog >= std::log(rng.rand())) {
            th(it) = pr.first;
            ph(it) = pr.second;
            update_inverse(cur, nxt, it, dr, temp);
            cur.copy_from(nxt, it);
            ++naccept;
        } else {
            nxt.copy_from(cur, it);
        }
        update_density(theta_mesh, th, density);
        it = (it + 1) % Nsys;
    }

    for (auto& d : density) d /= num_steps;
    save_csv("density_unprojected.csv", density);
    std::printf("sampling: acceptance=%.3f  (wrote density_unprojected.csv)\n",
                static_cast<double>(naccept) / num_steps);
    return 0;
}

File: examples/sampler_unprojected.jl
# # Sampling an unprojected composite-fermion state with Sherman-Morrison updates
#
# The *unprojected* CF wavefunction det[Y_{Q*,l,m}(Ωᵢ)] · ∏(uⱼvₖ-uₖvⱼ)^p is built from
# single-particle monopole-harmonic orbitals (`Ψunproj`). Because moving one particle
# changes only one column of the Slater determinant, the Slater inverse can be maintained
# by a rank-1 Sherman-Morrison update — turning the per-step cost from O(N³) (`logdet`) into
# O(N) for the acceptance ratio plus O(N²) for the inverse update. (This shortcut does NOT
# apply to the *projected* `Ψproj`/`Ψparton`, where a single move changes every column.)

using LinearAlgebra
LinearAlgebra.BLAS.set_num_threads(1)

using CFsOnSphere
using Random
const global RNG = Random.default_rng()

# Sample the unprojected CF ground state at ν = n/(2np+1) for `N` electrons.
function gibbs_sampler_unprojected(N::Int64, n::Int64, p::Int64, num_thermalization::Int64 = 5 * 10^5, num_steps::Int64 = 10^6)

    # Build the effective monopole strength and Λ-level occupation.
    Qstar, l_m_list = cf_ground_state_lm(N, n, p)
    system_size = length(l_m_list)

    # Two copies of the wavefunction: current (accepted) and next (proposed).
    Ψcurrent = Ψunproj(Qstar, p, system_size, l_m_list)
    Ψnext = Ψunproj(Qstar, p, system_size, l_m_list)

    # Random initial positions.
    θcurrent, ϕcurrent = rand_θ_ϕ_gen(RNG, system_size)
    θnext = copy(θcurrent)
    ϕnext = copy(ϕcurrent)

    # |Ψ|² = |det|² |Jastrow|²; gibbs_thermalization! maximises a real log-pdf.
    logpdf(ψ::Ψunproj) = 2.0 * real(logdet(ψ.slater_det) + ψ.jastrow_factor_log)

    # Phase 1 — thermalize and tune the step size σ with the generic driver (uses logdet).
    σ = pi / sqrt(12.0)
    sampling_iter, σ, _, _ = gibbs_thermalization!(RNG, Ψcurrent, Ψnext, θcurrent, ϕcurrent, θnext, ϕnext, σ, logpdf, num_thermalization)

    # Phase 2 — Sherman-Morrison-accelerated sampling. Maintain the inverse on Ψcurrent.
    initialize_inverse!(Ψcurrent)
    temp = zeros(ComplexF64, system_size)

    # Equal-area polar-angle density histogram.
    θmesh = map(x -> acos(x), LinRange(1.0, -1.0, 200))
    accumulated_density = zeros(Float64, length(θmesh) - 1)

    num_accepted = 0
    for monte_carlo_iter in 1:num_steps

        θnext[sampling_iter], ϕnext[sampling_iter] = proposal(RNG, θcurrent[sampling_iter], ϕcurrent[sampling_iter], σ)
        update_wavefunction!(Ψnext, θnext[sampling_iter], ϕnext[sampling_iter], sampling_iter)

        # O(N) determinant ratio from the maintained inverse and the new column.
        det_ratio = slater_det_ratio(Ψcurrent, Ψnext, sampling_iter)
        δlogpdf = 2.0 * real(log(det_ratio) + Ψnext.jastrow_factor_log - Ψcurrent.jastrow_factor_log)

        if δlogpdf >= log(rand(RNG))
            θcurrent[sampling_iter] = θnext[sampling_iter]
            ϕcurrent[sampling_iter] = ϕnext[sampling_iter]
            # O(N²) rank-1 inverse update, BEFORE copying the state across.
            update_inverse!(Ψcurrent, Ψnext, sampling_iter, det_ratio, temp)
            copy!(Ψcurrent, Ψnext, sampling_iter)
            num_accepted += 1
        else
            θnext[sampling_iter] = θcurrent[sampling_iter]
            ϕnext[sampling_iter] = ϕcurrent[sampling_iter]
            copy!(Ψnext, Ψcurrent, sampling_iter)
        end

        update_density!(θmesh, θcurrent, accumulated_density)

        sampling_iter = mod(sampling_iter, system_size) + 1
    end

    return θcurrent, ϕcurrent, accumulated_density ./ num_steps, num_accepted / num_steps
end

# A small smoke run (unprojected ν = 1/3, N = 9 in n = 3 filled Λ-levels).
# gibbs_sampler_unprojected(9, 3, 1, 10^4, 10^4)
